### PR TITLE
Provide fail-fast mechanism for starting docker container - fixes htt…

### PR DESCRIPTION
**Pull request by mnovak1**
…ps://github.com/jboss-eap-qe/eap-microprofile-test-suite/issues/28

### CODE REVIEW

File: tooling-docker/src/main/java/org/jboss/eap/qe/ts/common/docker/Docker.java [hunk 1]
```diff
@@ -77,7 +78,17 @@ protected void start() throws Exception {
             }
         });
 
-        containerReadyCondition.waitUntilReady(containerReadyTimeout);
+        long startTime = System.currentTimeMillis();
+        while (!(new CompletableFuture().supplyAsync(() -> containerReadyCondition.isReady())).get()) {
```
Comment (honza-kasik) [inline tooling-docker/src/main/java/org/jboss/eap/qe/ts/common/docker/Docker.java]: Just a curious question: Why is this async when there is a waiting for result?

Comment (mnovak1) [reply tooling-docker/src/main/java/org/jboss/eap/qe/ts/common/docker/Docker.java]: I did some tries to specify timeout on isReady() method so in case it's badly implemented, it does not hang forever and `containerReadyTimeout` will be applied however I did not finish it. I've changed the code so it's applied and added test for it.

Now when containerReadyCondition.isReady() will hang then after `containerReadyTimeout` new exception will be thrown.

Comment (mnovak1) [reply tooling-docker/src/main/java/org/jboss/eap/qe/ts/common/docker/Docker.java]: I'll mark this as resolved.

**Comment by mnovak1** [comment]
@rsvoboda thanks for spotting this, fixed and re-pushed.

**Comment by mnovak1** [comment]
@honza-kasik @zhantemirov @fabiobrz hi guys, could you take a look as well, please? thanks

File: tooling-docker/src/test/java/org/jboss/eap/qe/ts/common/docker/MalformedDockerConfigTest.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        thrown.expect(Exception.class);
+        thrown.expectMessage(startsWith("Starting of docker container using command: \"docker run --name"));
+        thrown.expectMessage(endsWith("failed. Check that provided command is correct."));
+
+        containerWithInvalidVersion.start();
+    }
+
+    @Test
+    public void testHangingReadyCondition() throws Exception {
+        Docker containerWithInvalidVersion = new Docker.Builder("wildfly", "jboss/wildfly:18.0.0.Final")
```
Comment (fabiobrz) [inline tooling-docker/src/test/java/org/jboss/eap/qe/ts/common/docker/MalformedDockerConfigTest.java]: I would change this from `containerWithInvalidVersion` to `hangingForReadyConditionContainer` or something similar.

File: tooling-docker/src/main/java/org/jboss/eap/qe/ts/common/docker/Docker.java [hunk 3]
```diff
@@ -77,7 +79,31 @@ protected void start() throws Exception {
...
         });
 
-        containerReadyCondition.waitUntilReady(containerReadyTimeout);
+        long startTime = System.currentTimeMillis();
+        while (!isContainerReady()) {
+            if (System.currentTimeMillis() - startTime > containerReadyTimeout) {
+                throw new Exception("Container was not ready in timeout : " + containerReadyTimeout + " ms");
+            }
+            // fail fast mechanism in case of malformed docker command, for example bad arguments, invalid format of port mapping, image version,...
+            if (!dockerRunProcess.isAlive() && dockerRunProcess.exitValue() != 0) {
```
Comment (fabiobrz) [inline tooling-docker/src/main/java/org/jboss/eap/qe/ts/common/docker/Docker.java]: Either one earlier check is to be done on docker process alone or I think I am getting a couple of green tests because of this - in conjunction with expected exception prefix [and suffix] as defined in test classes - even if I have docker daemon not running. 
WDYT?

File: tooling-docker/src/main/java/org/jboss/eap/qe/ts/common/docker/Docker.java [hunk 4]
```diff
@@ -77,7 +79,31 @@ protected void start() throws Exception {
...
+
+    private boolean isContainerReady() throws Exception {
+        CompletableFuture<Boolean> condition = new CompletableFuture().supplyAsync(() -> containerReadyCondition.isReady());
+        try {
+            return condition.get(containerReadyTimeout, TimeUnit.MILLISECONDS);
+        } catch (TimeoutException ex) {
+            // in case condition hangs interrupt it so there are no zombie threads
+            condition.cancel(true);
+
+            throw new Exception("Provided ContainerReadyCondition.isReady() method took longer than containerReadyTimeout: "
```
Comment (fabiobrz) [inline tooling-docker/src/main/java/org/jboss/eap/qe/ts/common/docker/Docker.java]: Could be providing a couple of strongly typed `Exception` sub-classes a good idea for refactoring? WDYT?

File: tooling-docker/src/test/java/org/jboss/eap/qe/ts/common/docker/MalformedDockerConfigTest.java (hunk 2 above)
Comment (mnovak1) [reply tooling-docker/src/test/java/org/jboss/eap/qe/ts/common/docker/MalformedDockerConfigTest.java]: what about `testContainerReadyTimeoutWithHangingCondition`?

File: tooling-docker/src/main/java/org/jboss/eap/qe/ts/common/docker/Docker.java (hunk 3 above)
Comment (mnovak1) [reply tooling-docker/src/main/java/org/jboss/eap/qe/ts/common/docker/Docker.java]: I'm not sure I understand. Do you want to update exception message that docker daemon might not be running?

File: tooling-docker/src/main/java/org/jboss/eap/qe/ts/common/docker/Docker.java (hunk 4 above)
Comment (mnovak1) [reply tooling-docker/src/main/java/org/jboss/eap/qe/ts/common/docker/Docker.java]: +1 I was too lazy to do so :-) I'll fix it.

File: tooling-docker/src/test/java/org/jboss/eap/qe/ts/common/docker/MalformedDockerConfigTest.java (hunk 2 above)
Comment (fabiobrz) [reply tooling-docker/src/test/java/org/jboss/eap/qe/ts/common/docker/MalformedDockerConfigTest.java]: Hi, it looks like the name for a _condition_, actually, rather than for something that represents a Docker instance. Using the same principle that you followed in the previous method (`containerWith*`) I'd have gone with something similar, like `containerWithHangingReadyCondition`. That said, I wouldn't want for us to trip over variable names too much, here.

Comment (mnovak1) [reply tooling-docker/src/test/java/org/jboss/eap/qe/ts/common/docker/MalformedDockerConfigTest.java]: `testContainerWithHangingReadyCondition` it is ;-)

File: tooling-docker/src/main/java/org/jboss/eap/qe/ts/common/docker/Docker.java (hunk 3 above)
Comment (fabiobrz) [reply tooling-docker/src/main/java/org/jboss/eap/qe/ts/common/docker/Docker.java]: It is not just about the exception message.
The message is the same - i.e. the expected one - that the `testFailFastWithMalformedDockerCommand` method is supposing to catch. 
So my concern is mainly about not getting a green bar from a fail-safe case method related to an unavailable image when instead tests have not been executed, actually - which instead I happen to be able to do.
That said I thought about a pre execution check for docker process to be running. WDYT?

Comment (mnovak1) [reply tooling-docker/src/main/java/org/jboss/eap/qe/ts/common/docker/Docker.java]: Thanks, now i get your point. I'll just rephrase it, we'll add check that docker command is present. This is because there is podman on rhel8 which does not have any docker daemon. 

I've added the check and re-pushed.

File: tooling-docker/src/test/java/org/jboss/eap/qe/ts/common/docker/MalformedDockerConfigTest.java (hunk 2 above)
Comment (fabiobrz) [reply tooling-docker/src/test/java/org/jboss/eap/qe/ts/common/docker/MalformedDockerConfigTest.java]: Now I get where we clashed... I meant the `Docker` variable name, while you addressed the method name. Anyway, if you don't want to change it, that's fine to me too.

File: tooling-docker/src/main/java/org/jboss/eap/qe/ts/common/docker/Docker.java [hunk 5]
```diff
@@ -77,7 +82,42 @@ protected void start() throws Exception {
...
+    }
+
+    private void checkDockerPresent() throws Exception {
+        Process dockerRunProcess = new ProcessBuilder()
+                .redirectErrorStream(true)
+                .command(new String[] {"docker", "info"})
+                .start();
+        dockerRunProcess.waitFor();
+        if (dockerRunProcess.exitValue() != 0)   {
+            throw new DockerException("Docker is not present on this machine. It must be installed for running tests with docker container.");
```
Comment (fabiobrz) [inline tooling-docker/src/main/java/org/jboss/eap/qe/ts/common/docker/Docker.java]: I'd rephrase 
`Docker is not present on this machine. It must be installed for running tests with docker container.`
to
`Docker is either not present or not installed on this machine. It must be installed and started up for executing tests with docker container.`

Comment (mnovak1) [reply tooling-docker/src/main/java/org/jboss/eap/qe/ts/common/docker/Docker.java]: thanks, fixed

**Comment by rsvoboda** [comment]
I have tried this on my home/slow machine with Windows and Docker Toolbox.
My machine is too slow to start 2 WF instances in time, what a great opportunity to see problems.

**1) error message**
```
[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 121.011 s <<< FAILURE! - in org.jboss.eap.qe.ts.common.docker.DockerTest
[ERROR] org.jboss.eap.qe.ts.common.docker.DockerTest  Time elapsed: 121.008 s  <<< ERROR!
org.jboss.eap.qe.ts.common.docker.DockerTimeoutException: Container was not ready in timeout : 120000 ms
        at org.jboss.eap.qe.ts.common.docker.Docker.start(Docker.java:88)
        at org.jboss.eap.qe.ts.common.docker.Docker.before(Docker.java:144)
```
Error message needs to mention container name for easy identification (was it wildfly-server-one or wildfly-server-two ?)
Container __TODO__ADD_NAME was not ready in timeout : 120000 ms


**2) running process for first WF after TS is finished**
```
$ docker ps
CONTAINER ID        IMAGE                        COMMAND                  CREATED             STATUS              PORTS                                              NAMES
220f0b63f5f6        jboss/wildfly:18.0.0.Final   "/opt/jboss/wildfly/"   5 minutes ago       Up 5 minutes        0.0.0.0:22222->8080/tcp, 0.0.0.0:22990->9990/tcp   1efb1c30-9130-496b-9e04-30107b4d85c9
```
Better cleanup needs to happen when the TestCase finishes.

In Quarkus QS there is stop command called before start to eliminate conflict with a leftover. I prefer to clean stuff right after the tests stops/fails, calling stop before running the container is workaround which could be applied if the "proper" way hits some limits
https://github.com/quarkusio/quarkus-quickstarts/blob/master/mqtt-quickstart/pom.xml#L97-L98

An option to skip such final cleanup for investigation can be handy too, could be tracked as a new issue.


**3) better names for started container**
1efb1c30-9130-496b-9e04-30107b4d85c9 is terrible thing for human investigation
We should have a human-readable name or at least prefix, at least something like wildfly-server-two-1efb1c30-9130-496b-9e04-30107b4d85c9 or something more stable / "rule base" like ${TEST_NAME}-wildfly-server-two
Could be tracked as a new issue.

Didn't look at code yet ;)

File: tooling-docker/src/main/java/org/jboss/eap/qe/ts/common/docker/Docker.java [hunk 6]
```diff
@@ -77,7 +82,43 @@ protected void start() throws Exception {
...
+            // in case condition hangs interrupt it so there are no zombie threads
+            condition.cancel(true);
+
+            throw new ContainerReadyConditionException("Provided ContainerReadyCondition.isReady() method took longer than containerReadyTimeout: "
+                    + containerReadyTimeout + " ms. Check it does not hang and does not take longer then containerReadyTimeout. " +
+                    "It's expected that ContainerReadyCondition.isReady() method is short lived (takes less than 1 second).", ex);
+        }
+    }
+
+    private void checkDockerPresent() throws Exception {
```
Comment (honza-kasik) [inline tooling-docker/src/main/java/org/jboss/eap/qe/ts/common/docker/Docker.java]: Shouldn't we check the output too? There might be `alias docker=podman` set up on the system (`podman-docker` package [does this for example when installed](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html-single/building_running_and_managing_containers/index#running-containers-without-docker_building-running-and-managing-containers)) and if we are looking specifically for the Docker, this might be an issue.

Comment (mnovak1) [reply tooling-docker/src/main/java/org/jboss/eap/qe/ts/common/docker/Docker.java]: @honza-kasik thanks, this check works with docker and podman as well. I've tried it on RHEL8.

**Comment by mnovak1** [comment]
@rsvoboda I've fixed 1) and 3). I'll take a look at 2) later today.

File: tooling-docker/src/main/java/org/jboss/eap/qe/ts/common/docker/Docker.java (hunk 6 above)
Comment (honza-kasik) [reply tooling-docker/src/main/java/org/jboss/eap/qe/ts/common/docker/Docker.java]: Thanks, that is not my concern. The issue is: Do we want to proceed with Podman?

Comment (mnovak1) [reply tooling-docker/src/main/java/org/jboss/eap/qe/ts/common/docker/Docker.java]: Podman is pushed on RHEL8 and at least for this moment we'll align with this. Current tooling depends on basic usage of `docker` command only so we're ok with having `alias docker=podman`.

File: tooling-docker/src/test/java/org/jboss/eap/qe/ts/common/docker/MalformedDockerConfigTest.java (hunk 2 above)
Comment (mnovak1) [reply tooling-docker/src/test/java/org/jboss/eap/qe/ts/common/docker/MalformedDockerConfigTest.java]: Right :-), I've changed it.

**Comment by mnovak1** [comment]
I've fixed 2) as well and added test for container ready timeout. I've also enhanced some tests with assert there is no left docker container.

**Comment by mnovak1** [comment]
All the above mentioned issues are fixed. Tests passed on RHEL7 and RHEL8. 

I would like to merge this as it has grown too big. In case there will be found new issues they'll be handled in separate PR.
